### PR TITLE
Add gRPC adhoc implementation.

### PR DESCRIPTION
**Pull request by apuchitnis**
Old PR: https://github.com/improbable-eng/kedge/pull/121

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -28,6 +28,8 @@ install-tools:
 	@go get -u golang.org/x/tools/cmd/goimports
 	@echo ">> fetching dep"
 	@go get -u github.com/golang/dep/cmd/dep
+	@go get -u github.com/mwitkow/go-proto-validators/protoc-gen-govalidators
```
Comment (Bplotka) [inline Makefile]: you don't need these. These packages are marked as required in kedge `Gopkg.toml`

Comment (Bplotka) [reply Makefile]: this means that any `make dep` will fetch them

File: pkg/kedge/grpc/director/adhoc/adhoc.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func NewStaticAddresser(rules []*kedge_config_common.Adhoc) *static {
+	return &static{rules: rules}
+}
+
+func (a *static) Address(hostString string) (string, error) {
+	hostName, port, err := common.ExtractHostPort(hostString)
+	if err != nil {
+		return "", grpc.Errorf(codes.InvalidArgument, "adhoc: malformed port number: %v", err)
```
Comment (Bplotka) [inline pkg/kedge/grpc/director/adhoc/adhoc.go]: Let's maybe we start using non deprecated package for gRPC errors? at least for new code? (: the new one is `status.Errorf`

Here and below

File: pkg/kedge/grpc/director/director.go [hunk 3]
```diff
@@ -8,27 +8,56 @@ import (
...
 )
 
 // New builds a StreamDirector based off a backend pool and a router.
-func New(pool backendpool.Pool, router router.Router) proxy.StreamDirector {
+func New(pool backendpool.Pool, adhocRouter common.Addresser, grpcRouter router.Router) proxy.StreamDirector {
 	return func(ctx context.Context, fullMethodName string) (context.Context, *grpc.ClientConn, error) {
-		beName, err := router.Route(ctx, fullMethodName)
+		beName, err := grpcRouter.Route(ctx, fullMethodName)
+
+		// Try adhoc router if RouteNotFound
```
Comment (Bplotka) [inline pkg/kedge/grpc/director/director.go]: missing trailing period in comment

File: pkg/kedge/grpc/backendpool/backend.go [hunk 4]
```diff
@@ -65,7 +65,7 @@ func (b *backend) Conn() (*grpc.ClientConn, error) {
 	b.target = target
 	b.resolver = resolver
 
-	cc, err = buildClientConn(b.config, target, resolver)
+	cc, err = BuildClientConn(b.config, target, resolver)
```
Comment (Bplotka) [inline pkg/kedge/grpc/backendpool/backend.go]: Why this change? We seem to not use BuildClientConn cross package.

File: pkg/kedge/grpc/director/director.go [hunk 5]
```diff
@@ -8,27 +8,56 @@ import (
...
+		if err == router.ErrRouteNotFound {
+			ipPort, err := adhocRouter.Address(metautils.ExtractIncoming(ctx).Get(":authority"))
+			if err != nil {
+				return ctx, nil, errors.Wrap(err, "failed to proxy request using adhoc router")
```
Comment (Bplotka) [inline pkg/kedge/grpc/director/director.go]: this is not always true message. Most likely we will get `router.ErrRouteNotFound` if no adhoc is specified. 

I would actually vote for not wrapping this error. It is already nicely wrapped in `Address` (`adhoc` prefix)

File: pkg/kedge/grpc/director/director.go [hunk 6]
```diff
@@ -8,27 +8,56 @@ import (
...
+			cc, err := grpc.Dial(ipPort, opts...)
+			if err != nil {
+				return ctx, nil, errors.Wrapf(err, "failed to dial to adhoc backend %v", ipPort)
+			}
+
+			go func() {
+				<-ctx.Done()
+				cc.Close()
+			}()
+
```
Comment (Bplotka) [inline pkg/kedge/grpc/director/director.go]: can we add here some log tag? for example:

`grpc_ctxtags.Extract(ctx).Set("grpc.proxy.adhoc", ipPort)`

File: pkg/kedge/grpc/director/director.go [hunk 7]
```diff
@@ -8,27 +8,56 @@ import (
...
+			return grpcutils.CloneIncomingToOutgoingMD(ctx), cc, err
+		}
+
+		// Return all other errors
```
Comment (Bplotka) [inline pkg/kedge/grpc/director/director.go]: missing trailing period

File: pkg/kedge/grpc/director/director.go [hunk 8]
```diff
@@ -8,27 +8,56 @@ import (
...
 		if err != nil {
 			return ctx, nil, err
 		}
 
 		grpc_ctxtags.Extract(ctx).Set("grpc.proxy.backend", beName)
-		ctx = grpcutils.CloneIncomingToOutgoingMD(ctx)
-
 		cc, err := pool.Conn(beName)
-		return ctx, cc, err
+		return grpcutils.CloneIncomingToOutgoingMD(ctx), cc, err
```
Comment (Bplotka) [inline pkg/kedge/grpc/director/director.go]: nice

File: pkg/kedge/grpc/integration_test.go [hunk 9]
```diff
@@ -124,7 +144,7 @@ type localBackends struct {
 }
 
 func (l *localBackends) addServer(t *testing.T, serverOpt ...grpc.ServerOption) {
-	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	listener, err := net.Listen("tcp", "127.0.0.1:")
```
Comment (Bplotka) [inline pkg/kedge/grpc/integration_test.go]: just curious does it change anything?

File: pkg/kedge/grpc/integration_test.go [hunk 10]
```diff
@@ -238,12 +259,18 @@ func (s *BackendPoolIntegrationTestSuite) SetupSuite() {
 	// Make ourselves the resolver for SRV for our backends. See Lookup function.
 	s.originalSrvResolver = srvresolver.ParentSrvResolver
 	srvresolver.ParentSrvResolver = s
+
+	// Make ourselves the A resolver for backends for the Addresser.
+	s.originalAResolver = common.DefaultALookup
```
Comment (Bplotka) [inline pkg/kedge/grpc/integration_test.go]: you can do `s.originalAResolver, common.DefaultALookup = common.DefaultALookup, lookupAddr` but not a blocker

File: pkg/kedge/grpc/integration_test.go [hunk 11]
```diff
@@ -362,7 +397,23 @@ func (s *BackendPoolIntegrationTestSuite) TestCallToNonSecureBackendLoadBalances
 
 func (s *BackendPoolIntegrationTestSuite) TestCallToUnknownRouteCausesError() {
 	err := grpc.Invoke(s.SimpleCtx(), "/bad.route.doesnt.exist/Method", &unknownResponse{}, &unknownResponse{}, s.proxyConn)
-	require.EqualError(s.T(), err, "rpc error: code = Unimplemented desc = unknown route to service", "no error on simple call")
+	require.EqualError(s.T(), err, "rpc error: code = Unknown desc = failed to proxy request using adhoc router: rpc error: code = Unimplemented desc = unknown route to service", "no error on simple call")
```
Comment (Bplotka) [inline pkg/kedge/grpc/integration_test.go]: Again, error should stay the same IMO for this case (and that would consistent with http)

File: pkg/kedge/grpc/integration_test.go [hunk 12]
```diff
@@ -362,7 +397,23 @@ func (s *BackendPoolIntegrationTestSuite) TestCallToNonSecureBackendLoadBalances
...
+}
+
+func (s *BackendPoolIntegrationTestSuite) TestSuccessOverForwardProxy_ApuTesting() {
```
Comment (Bplotka) [inline pkg/kedge/grpc/integration_test.go]: hm.. you want to leave `ApuTesting` suffix?  (: Like a digital signature ^^

File: Makefile (hunk 1 above)
Comment (apuchitnis) [reply Makefile]: ahh, I didn't run `dep ensure` when I first started working on the project it seems (or my GOBIN wasn't sourced). will remove

File: pkg/kedge/common/common.go [hunk 13]
```diff
@@ -89,3 +87,11 @@ func PortAllowed(port int, portRule *pb.Adhoc_Port) bool {
 	}
 	return false
 }
+
+func ResolveHost(hostStr string) (string, error) {
+	addrs, err := DefaultALookup(hostStr)
+	if err != nil {
+		return "", err
+	}
+	return addrs[0], nil
```
Comment (adamhosier) [inline pkg/kedge/common/common.go]: is there ever a case where `err == nil` but `len(addrs) == 0`? not to familiar with the net package

Comment (apuchitnis) [reply pkg/kedge/common/common.go]: good point, looking at other usages in other packages this can happen. will handle that case

Comment (apuchitnis) [reply pkg/kedge/common/common.go]: cc @Bplotka

Comment (Bplotka) [reply pkg/kedge/common/common.go]: This is very good point - I don't know about net package but this is interface so we need to expect anything!

File: pkg/kedge/common/common.go [hunk 14]
```diff
@@ -89,3 +87,14 @@ func PortAllowed(port int, portRule *pb.Adhoc_Port) bool {
...
 	return false
 }
+
+func ResolveHost(hostStr string) (string, error) {
+	addrs, err := DefaultALookup(hostStr)
+	if err != nil {
+		return "", err
+	}
+	if len(addrs) == 0 {
+		return "", err
```
Comment (Bplotka) [inline pkg/kedge/common/common.go]: mhmhm that would trigger panics too potentially. Let's error here properly (err will be always nil here)

File: pkg/kedge/common/common.go [hunk 15]
```diff
@@ -8,6 +6,9 @@ import (
 	"strings"
 	"sync"
```
Comment (Bplotka) [inline pkg/kedge/common/common.go]: let's remove that space & run  make format (yea goimports is sometimes silly)

File: pkg/kedge/common/common.go [hunk 16]
```diff
@@ -89,3 +90,14 @@ func PortAllowed(port int, portRule *pb.Adhoc_Port) bool {
...
 	return false
 }
+
+func ResolveHost(hostStr string) (string, error) {
+	addrs, err := DefaultALookup(hostStr)
+	if err != nil {
+		return "", err
+	}
+	if len(addrs) == 0 {
+		return "", errors.New(fmt.Sprintf("did not find any A records for host %v", hostStr))
```
Comment (Bplotka) [inline pkg/kedge/common/common.go]: mhmhm there is `errors.Errorf` for that man (:
